cpp/src/arrow/compute/exec/hash_join_benchmark.cc

```diff
@@ -0,0 +1,406 @@
+}
+
+template <typename... Args>
+static void BM_HashJoinBasic_Selectivity(benchmark::State& st,
+                                         std::vector<std::shared_ptr<DataType>> key_types,
+                                         Args&&...) {
+  BenchmarkSettings settings;
+  settings.selectivity = static_cast<double>(st.range(0)) / 100.0;
+
+  settings.num_build_batches = static_cast<int>(st.range(1));
+  settings.num_probe_batches = settings.num_build_batches;
+  settings.key_types = std::move(key_types);
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+template <typename... Args>
+static void BM_HashJoinBasic_JoinType(benchmark::State& st, JoinType join_type,
+                                      Args&&...) {
+  BenchmarkSettings settings;
+  settings.selectivity = static_cast<double>(st.range(0)) / 100.0;
+
+  settings.num_build_batches = static_cast<int>(st.range(1));
+  settings.num_probe_batches = settings.num_build_batches;
+  settings.join_type = join_type;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+static void BM_HashJoinBasic_MatchesPerRow(benchmark::State& st) {
+  BenchmarkSettings settings;
+  settings.cardinality = 1.0 / static_cast<double>(st.range(0));
+
+  settings.num_build_batches = static_cast<int>(st.range(1));
+  settings.num_probe_batches = settings.num_probe_batches;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+static void BM_HashJoinBasic_PayloadSize(benchmark::State& st) {
+  BenchmarkSettings settings;
+  int32_t payload_size = static_cast<int32_t>(st.range(0));
+  settings.probe_payload_types = {fixed_size_binary(payload_size)};
+  settings.cardinality = 1.0 / static_cast<double>(st.range(1));
+
+  settings.num_build_batches = static_cast<int>(st.range(2));
+  settings.num_probe_batches = settings.num_probe_batches;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+static void BM_HashJoinBasic_ProbeParallelism(benchmark::State& st) {
+  BenchmarkSettings settings;
+  settings.num_threads = static_cast<int>(st.range(0));
+  settings.num_build_batches = static_cast<int>(st.range(1));
+  settings.num_probe_batches = settings.num_build_batches * 8;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+static void BM_HashJoinBasic_BuildParallelism(benchmark::State& st) {
+  BenchmarkSettings settings;
+  settings.num_threads = static_cast<int>(st.range(0));
+  settings.num_build_batches = static_cast<int>(st.range(1));
+  settings.num_probe_batches = settings.num_threads;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+static void BM_HashJoinBasic_NullPercentage(benchmark::State& st) {
+  BenchmarkSettings settings;
+  settings.null_percentage = static_cast<double>(st.range(0)) / 100.0;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+std::vector<std::string> keytypes_argnames = {"Hashtable krows"};
+BENCHMARK_CAPTURE(BM_HashJoinBasic_KeyTypes, "{int32}", {int32()})
+    ->ArgNames(keytypes_argnames)
+    ->RangeMultiplier(4)
+    ->Range(1, 4 * 1024);
```

Comment:
If I understand correctly, this is generating `{1, 4, 16, 64, 256, 1024, 4096}`. You could cut down on the number of generated tests by making the multiplier 8 or 16, IMHO, and you probably wouldn't lose much in performance insight?

Reply:
I tested it, and it only shaves 1.5 min off the runtime. But I will unify it among all tests and set it to 8.